Forta-Bots-Challenge-2/src/agent.spec.ts

```diff
@@ -0,0 +1,216 @@
+import { createTestFinding } from "./findings";
+import { BigNumber } from "ethers";
+import { PoolData } from "./uniswap.fetcher";
+
+// format: [blockNumber, token0, token1, fee]
+const TEST_DATA: [number, string, string, BigNumber][] = [
+  [10, createAddress("0xaa1"), createAddress("0xbb1"), BigNumber.from(1)],
+  [20, createAddress("0xaa2"), createAddress("0xbb2"), BigNumber.from(2)],
+  [30, createAddress("0xaa3"), createAddress("0xbb3"), BigNumber.from(3)],
+];
+
+// format: [sender,  recipient,  amount0, amount1, sqrtPriceX96, liquidity, tick]
+const TEST_LOGS: [string, string, BigNumber, BigNumber, BigNumber, BigNumber, BigNumber][] = [
+  [
+    createAddress("0xae1"),
+    createAddress("0xbe2"),
+    BigNumber.from(10),
+    BigNumber.from(10),
+    BigNumber.from(1),
+    BigNumber.from(100),
+    BigNumber.from(1),
+  ],
+  [
+    createAddress("0xae2"),
+    createAddress("0xce3"),
+    BigNumber.from(5),
+    BigNumber.from(15),
+    BigNumber.from(2),
+    BigNumber.from(50),
+    BigNumber.from(2),
+  ],
+  [
+    createAddress("0xab6"),
+    createAddress("0xcb5"),
+    BigNumber.from(50),
+    BigNumber.from(1),
+    BigNumber.from(20),
+    BigNumber.from(500),
+    BigNumber.from(3),
+  ],
+];
+
+const TEST_POOL_DATA_0: PoolData = {
+  isUniswapPool: true,
+  token0: "TKN0",
+  token1: "TKN1",
+  fee: 1,
+};
+
+const TEST_POOL_DATA_1: PoolData = {
+  isUniswapPool: true,
+  token0: "TKN0",
+  token1: "TKN1",
+  fee: 2,
+};
+
+const TEST_POOL_DATA_2: PoolData = {
+  isUniswapPool: true,
+  token0: "TKN0",
+  token1: "TKN1",
+  fee: 3,
+};
+
+describe("uniswap v3 swap detection", () => {
+  let handleTransaction: HandleTransaction;
+  let initialize: Initialize;
+  let mockTxEvent = new TestTransactionEvent();
+  const mockProvider: MockEthersProvider = new MockEthersProvider();
+  const mockUniswapFactory = createAddress("0x01");
+  const mockInitCodeHash = keccak256(createAddress("0x02"));
+  const ERC20_IFACE = new Interface(ERC20_ABI);
+  const UNISWAP_POOL_IFACE = new Interface(UNISWAP_POOL_ABI);
+  const INCORRECT_IFACE = new Interface(["function foo() external view returns (address)"]);
+
+  const getPoolAddress = (
+    factoryAddress: string,
+    initCodeHash: string,
+    saltValues: [string, string, BigNumber]
+  ): string => {
+    const encoded = defaultAbiCoder.encode(["address", "address", "uint24"], saltValues);
+    const salt = solidityKeccak256(["bytes"], [encoded]);
+    return getCreate2Address(factoryAddress, salt, initCodeHash);
+  };
+
+  const mockPoolCalls = (poolAddress: string, block: number, iface: Interface, calls: [string, any][]) => {
+    for (let [call, output] of calls) {
+      mockProvider.addCallTo(poolAddress, block, iface, call, {
+        inputs: [],
+        outputs: [output],
+      });
+    }
+  };
+
+  const mockERC20Calls = (token0: string, token1: string, block: number) => {
```

Comment:
Similar comment to fetcher test suite re: these sort of helper functions being outside the `describe()` block